### PR TITLE
Appointment endpoints

### PR DESCRIPTION
**Pull request by MariusAchitei**

### CODE REVIEW

File: src/controllers/appointments.controller.ts [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+import User from "../models/user";
+import Appointment from "../models/appointment";
+// import Result from "../models/results";
+
+export const getAppointments = async (req: Request, res: Response) => {
+  const { id } = res.locals.user;
+
+  const currentUser = await User.findById(id);
+  if (currentUser == null) {
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: use === insead of ==

File: src/controllers/appointments.controller.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
+import { Request, Response } from "express";
+const express = require("express");
+
+import User from "../models/user";
+import Appointment from "../models/appointment";
+// import Result from "../models/results";
+
+export const getAppointments = async (req: Request, res: Response) => {
+  const { id } = res.locals.user;
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: you take the id from the response object?

File: src/controllers/appointments.controller.ts [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+  const currentUser = await User.findById(id);
+  if (currentUser == null) {
+    res.status(404);
+    res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: use the methods like this: res.status(404).send(...)

File: src/controllers/appointments.controller.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    });
+  }
+
+  const appointments = await Appointment.find({
+    $or: [{ patient: id }, { doctor: id }],
+  });
+
+  console.log(appointments);
+  res.status(200);
+  res.send({ succes: true, message: appointments });
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again, call each method by the returned object of the last one: res.status(200).send(...)

File: src/controllers/appointments.controller.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+};
+
+export const postAppointment = async (req: Request, res: Response) => {
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: postAppointment -> addAppointment

File: src/controllers/appointments.controller.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  const { id } = res.locals.user;
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: you take the id from the response object?

File: src/controllers/appointments.controller.ts [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+  const newAppointent = new Appointment({
+    patient: currentUser,
+    doctor,
+    date,
+    type,
+    isActive,
+  });
+  newAppointent.save();
+  console.log(newAppointent);
+  res.send({ succes: true, message: newAppointent });
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: add a status to the response

File: src/controllers/appointments.controller.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+};
+
+export const getAppointment = async (req: Request, res: Response) => {
+  const { id } = res.locals.user;
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again with the response object

File: src/controllers/appointments.controller.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+  const { appId } = req.params;
+  const currentApp = await Appointment.findById(appId);
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: currentApp -> currentAppointment

File: src/controllers/appointments.controller.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+  if (currentApp == null) {
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: use === instead of ==

File: src/controllers/appointments.controller.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    res.status(404);
+    res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again, format them as one after the other

File: src/controllers/appointments.controller.ts [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+      succes: false,
+      message: "The given appointment ID was not found",
+    });
+  }
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: use === insead of ==
And what is this line of code doing exactly?

File: src/controllers/appointments.controller.ts [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    {
+      console.log("Refused the getting of an appointment");
+      console.log("Cause: unauthorized");
+      res.status(401);
+      res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+    res.send({
+      succes: false,
+      message: "The given appointment ID was not found",
+    });
+  }
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
+    {
+      console.log("Refused the getting of an appointment");
+      console.log("Cause: unauthorized");
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: The user should not get here without a token, so we assume the if the client got to this step he is already authorized

File: src/controllers/appointments.controller.ts [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+      res.status(401);
+      res.send({
+        succes: false,
+        message:
+          "Refused the getting of an appointment\n" + "Cause: unauthorized",
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: just "Unauthorized"

File: src/controllers/appointments.controller.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+      });
+      return;
+    }
+  }
+  res.status(200);
+  res.send({ succes: true, message: currentApp });
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    res.send({
+      succes: false,
+      message: "The given appointment ID was not found",
+    });
+  }
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
+    {
+      console.log("Refused the editing of an appointment");
+      console.log("Cause: unauthorized");
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: just 'Unauthoried'

File: src/controllers/appointments.controller.ts [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+  const currentApp = await Appointment.findById(appId);
+  if (currentApp == null) {
+    res.status(404);
+    res.send({
+      succes: false,
+      message: "The given appointment ID was not found",
+    });
+  }
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: === instead of ==

File: src/controllers/appointments.controller.ts [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+      });
+      return;
+    }
+  }
+  res.status(200);
+  res.send({ succes: true, message: currentApp });
+};
+
+export const deActivateAppointment = async (req: Request, res: Response) => {
+  const { id } = res.locals.user;
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: from response?

File: src/controllers/appointments.controller.ts [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+  const { appId } = req.params;
+  const currentApp = await Appointment.findById(appId);
+  if (currentApp == null) {
+    res.status(404);
+    res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+      message: "The given appointment ID was not found",
+    });
+  }
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
+    {
+      console.log("Refused the editing of an appointment");
+      console.log("Cause: unauthorized");
+      res.status(401);
+      res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+        message:
+          "Refused the editing of an appointment\n" + "Cause: unauthorized",
+      });
+      return;
+    }
+  }
+  currentApp.isActive = false;
+  await currentApp.save();
+  res.status(200);
+  res.send({ succes: true, message: "Deactivated appointment" });
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+      res.status(401);
+      res.send({
+        succes: false,
+        message:
+          "Refused the editing of an appointment\n" + "Cause: unauthorized",
+      });
+      return;
+    }
+  }
+  currentApp.isActive = false;
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 24]
```diff
@@ -0,0 +1,139 @@
...
+  await currentApp.save();
+  res.status(200);
+  res.send({ succes: true, message: "Deactivated appointment" });
+};
+
+export const deleteAppointment = async (req: Request, res: Response) => {
+  const { id } = res.locals.user;
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+  const { appId } = req.params;
+  const currentApp = await Appointment.findById(appId);
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+  if (currentApp == null) {
+    res.status(404);
+    res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+      succes: false,
+      message: "The given appointment ID was not found",
+    });
+  }
+  //   console.log(`DOCTOR: [${currentApp.doctor}]`);
+  //   console.log(id);
+  //   console.log(`PATIENT: [${currentApp.patient}]`);
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 28]
```diff
@@ -0,0 +1,139 @@
...
+    {
+      console.log("Refused the deletion of an appointment");
+      console.log("Cause: unauthorized");
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 29]
```diff
@@ -0,0 +1,139 @@
...
+      res.status(401);
+      res.send({
+        succes: false,
+        message:
+          "Refused the deletion of an appointment\n" + "Cause: unauthorized",
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 30]
```diff
@@ -0,0 +1,139 @@
...
+  //   console.log(`DOCTOR: [${currentApp.doctor}]`);
+  //   console.log(id);
+  //   console.log(`PATIENT: [${currentApp.patient}]`);
+
+  if (!currentApp.patient == id && currentApp.patient == id) {
+    {
+      console.log("Refused the deletion of an appointment");
+      console.log("Cause: unauthorized");
+      res.status(401);
+      res.send({
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/controllers/appointments.controller.ts [hunk 31]
```diff
@@ -0,0 +1,139 @@
...
+        succes: false,
+        message:
+          "Refused the deletion of an appointment\n" + "Cause: unauthorized",
+      });
+      return;
+    }
+  }
+  Appointment.findByIdAndRemove(appId);
+  res.status(200);
+  res.send({ succes: true, message: "Deleted appointment" });
```
Comment (LucaStefan112) [inline src/controllers/appointments.controller.ts]: again

File: src/index.ts [hunk 32]
```diff
@@ -1,16 +1,20 @@
-import mongoose from "mongoose";
 import { dbConnect } from "./db/connection";
 const express = require("express");
 const app = express();
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
 
-app.use(express.json())
-app.use(express.urlencoded({ extended: true }))
+import { limiter } from "./middlewares/limiter";
```
Comment (LucaStefan112) [inline src/index.ts]: imports to the top

File: src/index.ts [hunk 33]
```diff
@@ -1,16 +1,20 @@
...
+import appointmentRouter from "./routes/appointment";
```
Comment (LucaStefan112) [inline src/index.ts]: imports to the top

File: src/middlewares/validation.ts [hunk 34]
```diff
@@ -0,0 +1,13 @@
+import { Express, Request, Response, NextFunction } from "express";
+
+export const validateToken = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const { token } = req.params;
+  console.log(token);
```
Comment (LucaStefan112) [inline src/middlewares/validation.ts]: remove log

File: src/middlewares/validation.ts [hunk 35]
```diff
@@ -0,0 +1,13 @@
+import { Express, Request, Response, NextFunction } from "express";
+
+export const validateToken = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const { token } = req.params;
```
Comment (LucaStefan112) [inline src/middlewares/validation.ts]: the token should be taken from the cookie

File: src/models/appointment.ts [hunk 36]
```diff
@@ -3,46 +3,49 @@ import mongoose from "mongoose";
 import { AppointmentTypes } from "../helper/enums";
 
 export interface IAppointment {
-    patient: string;
-    doctor: string;
-    date: Date;
-    type: string;
-    meetingLink: string;
-    isActive: boolean;
+  patient: string;
```
Comment (LucaStefan112) [inline src/models/appointment.ts]: implement a patient interface

File: src/models/appointment.ts [hunk 37]
```diff
@@ -3,46 +3,49 @@ import mongoose from "mongoose";
...
+  doctor: string;
```
Comment (LucaStefan112) [inline src/models/appointment.ts:8]: implement a doctor interface

File: src/routes/appointment.ts [hunk 38]
```diff
@@ -0,0 +1,28 @@
+const express = require("express");
+const appointmentRouter = express.Router();
+
+import { validateToken } from "../middlewares/validation";
```
Comment (LucaStefan112) [inline src/routes/appointment.ts]: imports at the top

File: src/routes/appointment.ts [hunk 39]
```diff
@@ -0,0 +1,28 @@
...
+
+import {
```
Comment (LucaStefan112) [inline src/routes/appointment.ts:2]: at the top

File: src/routes/appointment.ts [hunk 40]
```diff
@@ -0,0 +1,28 @@
...
+  getAppointments,
+  postAppointment,
+  deleteAppointment,
+  deActivateAppointment,
+  getAppointment,
+} from "../controllers/appointments.controller";
+
+appointmentRouter.use("/:token", validateToken);
```
Comment (LucaStefan112) [inline src/routes/appointment.ts]: validate token should be used by the app object

**Review by LucaStefan112 — CHANGES_REQUESTED** [review]

File: src/controllers/appointments.controller.ts (hunk 2 above)
Comment (MariusAchitei) [reply src/controllers/appointments.controller.ts]: Asa nu mi da eroare ca la Omer de aia automata, e il setez cand se verifica tokenu si e mai simplu :)))

**Review by LucaStefan112 — APPROVED** [review]

**Pull request merged**